HPCC-34245 Unit test the thread pool class

streeterd opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

streeterd committed:
HPCC-34245 Unit test the thread pool class

Unit tests added for IThreadPool class.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd commented:
### Note
This unit test has 3 tests that will fail due to the HPCC-34258 issue.
One test is commented out as it is dependent upon HPCC-34238 being merged.

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        // This should block for approximately 100ms (start delay) then throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4", 100); // timeout after 100ms
```
arguably, if there are 2 test functions, 1 for tightly bound, 1 for throttled testing, then this test belongs more in the tightly bound realm

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+            StringBuffer msg;
+            e->errorMessage(msg);
+            CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPool", msg.str());
+            e->Release();
```
trivial, but should really get the msg, and release the IException, before any potential assert failure.

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        }
+        catch (IException *e)
+        {
+            StringBuffer msg("Exception encountered: ");
```
should this exception handling be same (common code) exception handling as the previous one? i.e. check message and time >= X

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        auto longThrottleEndTime = std::chrono::high_resolution_clock::now();
+        auto longThrottleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(longThrottleEndTime - longThrottleStartTime);
+
+        CPPUNIT_ASSERT(longThrottleDuration.count() >= 450 && longThrottleDuration.count() <= 600);
```
if how can it be 450 - 499 ?

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        CPPUNIT_ASSERT(!pool->waitAvailable(100)); // Should timeout since pool is full
+
+        // Wait for threads to finish
+        std::this_thread::sleep_for(std::chrono::seconds(3));
```
see other comments. Should use signals to know

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+
+        // Test: Pool was above, fell below, and is now at defaultmax again
+        pool->start(&lifespan2seconds, "Thread7"); // Now at capacity (3 threads)
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```
not really sure what this is for - to give 1ms time for them to begin? too risky to guarantee this way.
See other comments, need to use semaphores really to know.

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+            auto reThrottleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(reThrottleEndTime - reThrottleStartTime);
+
+            CPPUNIT_ASSERT(reThrottleDuration.count() >= 150 && reThrottleDuration.count() <= 300);
+            e->Release();
```
should release before

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+            auto reThrottleEndTime = std::chrono::high_resolution_clock::now();
+            auto reThrottleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(reThrottleEndTime - reThrottleStartTime);
+
+            CPPUNIT_ASSERT(reThrottleDuration.count() >= 150 && reThrottleDuration.count() <= 300);
```
why >= 150?

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+                pool->start(&lifespan1second, threadName.str());
+            }
+
+            std::this_thread::sleep_for(std::chrono::seconds(2));
```
would be good to add a comment why expected to be <2 secs (100 immediate, 100 delayed for.., total time expected to be ~..)

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+
+            for (int i = 201; i <= 400; i++)
+            {
+                std::thread asyncStart([&i, this, &exceptionCount, &totalStartDuration]()
```
formatting: it's more readable if you put the lambda on it's own line, e.g.:
```
auto threadFunc = [&i, this, &exceptionCount, &totalStartDuration]()
{
...
};
std::thread asyncStart(threadFunc);
```

jakesmith replied:
trivial: slightly misleading to pass 'i' in by reference (is something altering it? no)

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+                std::thread asyncStart([&i, this, &exceptionCount, &totalStartDuration]()
+                                       {
+                                                StringBuffer threadName;
+                                                threadName.appendf("Thread%d", i);
```
trivial: would be better if 'i' was an unsigned and %u used

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+                                                auto startDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterDelayedStart - beforeDelayedStart);
+                                                totalStartDuration += startDuration.count(); });
+
+                asyncStart.detach();
```
this probably not a good idea, i.e. leaving it dangling. Should track and join them all.

Although saying that, I am not convinced it is worth creating 200 std::threads, that each call pool->start ..
It might be worth redoing a another 200 start (as you did on lines 355 -> 360, to make sure limits and throttling are still consistent though..

streeterd replied:
Fixed in 210625692eacb9a827670f00eb0611004b0ef94b

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        auto endTime = std::chrono::high_resolution_clock::now();
+        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
+
+        CPPUNIT_ASSERT(duration.count() >= 180 && duration.count() <= 250);
```
why 180? and is 50ms margin enough (should us esignals and failure on longer than expected signalling )

jakesmith commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+        CPPUNIT_ASSERT(duration.count() >= 180 && duration.count() <= 250);
+
+        // Test: waitAvailable should return true immediately when slot becomes available
+        std::this_thread::sleep_for(std::chrono::seconds(5));
```
I think should aim for no thread in these tests to run for > ~ 1s.

streeterd replied:
Fixed in 210625692eacb9a827670f00eb0611004b0ef94b

jakesmith reviewed (changes requested):
@streeterd - I think # of tetsts (and size of code) should reduced, employ helpers for code reuse, and use semaphore patterns to check when threads are completes (rather than timings).

streeterd committed:
1.Use CSimpleInterface instead of CInterface
2.Rename cpp to jthreadpooltests.cpp
3.Add override to setUp() and tearDown()
4.Initialize members, even if reinitialized later
5.Standard is initialization
6.Use a semaphore, signal and wait instead of sleep for test timing
7.Ensure test name reflects the tests within
8.Remove redundant tests and combine to reduce the number of tests
9.Release the IException before any potential CPP Assert failures
10.Add helper functions to reduce code overhead and duplication
11.Do not use async start for 200 new threads instead use loop
12.Recode to ensure no thread has a runtime of more than 1 second
13.Remove test dependent upon HPCC-34238 as it can be added by
HPCC-34238
14.Use constexpr instead of primitive contant member variables
15.Move in other jthread tests e.g. ThreadedPersistStressTest
16.Incorporate testThrottledThreadPoolWithSpecifiedDelay into
testThrottledThreadPoolWithDefaultDelay

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+class TestPooledThread : public CInterface, implements IPooledThread
+{
+public:
+    IMPLEMENT_IINTERFACE;
```
Fixed in 210625692eacb9a827670f00eb0611004b0ef94b

streeterd commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+class TestThreadFactory : public CInterface, implements IThreadFactory
+{
+public:
+    IMPLEMENT_IINTERFACE;
```
Fixed in 210625692eacb9a827670f00eb0611004b0ef94b

streeterd commented on testing/unittests/jthreadtests.cpp:
```diff
@@ -0,0 +1,591 @@
+
+private:
+    std::atomic<unsigned> threadStartedCount;
+    std::atomic<unsigned> threadCompletedCount;
```
Fixed in 210625692eacb9a827670f00eb0611004b0ef94b

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        unsigned thread4ExceptionCount = 0;
+        attemptThreadStartWithTimeout(&params, milliseconds1000, "ThrottledThreadPoolWithDefaultDelay", thread4ExceptionCount, false);
+        duration = measureThreadStartDuration(beforeStart, 1);
+        CPPUNIT_ASSERT(duration.count() >= 290);
```
magic number, what is 290? I think should be:
 min = 500 (initial 2 threads) - 200 (thread3 delay) = 300? 

.. so I'm actually surprised this doesn't fail 10ms margin though is tight..

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        CPPUNIT_ASSERT(threadCompletedCount.load() >= 1);
+        CPPUNIT_ASSERT(pool->runningCount() >= 1);
+
+        // Test Thread Pool default start delay
```
not sure what this is testing? 
there should be a free slot I think...

maybe comment only misleading/needs clarifying.

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        CPPUNIT_ASSERT(duration.count() < 100);
+        CPPUNIT_ASSERT(pool->runningCount() >= 2);
+
+        // Thread6 should start up to the Thread Pool default start delay of 1 second
```
should start up *after* the thread pool start delay of 1 second ?

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        beforeStart = std::chrono::high_resolution_clock::now();
+        pool->start(&params, params.name); // no exception should be thrown as no timeout is specified on the start
+        duration = measureThreadStartDuration(beforeStart, 1);
+        CPPUNIT_ASSERT(duration.count() >= 749 && duration.count() < 1000);
```
could be less than 749 couldn't it? needs an comment explaining logic of expectation.

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+                                   this->pool->start(&asyncParams, asyncParams.name, INFINITE); });
+        asyncStart.detach();
+        duration = measureThreadStartDuration(beforeStart, 1);
+        CPPUNIT_ASSERT(duration.count() > 990);
```
these numbers need rationalizing and clarifying the logic/expections.

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        CPPUNIT_ASSERT(duration.count() > 990);
+        CPPUNIT_ASSERT_EQUAL(8U, threadStartedCount.load());
+        CPPUNIT_ASSERT(threadCompletedCount.load() >= 1);
+        CPPUNIT_ASSERT(pool->runningCount() >= 1);
```
>=1 ? are numbers non-determinate at this stage?

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+    void testThrottledThreadPoolWithSpecifiedDelay()
+    {
+        // Test: Expected throttling delays are introduced when pool is at capacity
+        pool.setown(createThreadPool("ThrottledThreadPool", factory, true, nullptr, 3, 500));
```
stopped reviewing this test beyond this line, but it sounds like the same as previous except with .5 s vs default 1s ?
If so, it is should at least be combined into a single helper, but on the face of it, it doesn't sounds like it's testing anything very different

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+            {
+                threadName.appendf("Thread%u", threadId);
+                params = {lifespan1000milliseconds, threadName.str()};
+                attemptThreadStartWithTimeout(&params, 100, "TestPoolWithFastThreadCompletion", exceptionCount,
```
re. earlier comment about attemptThreadStartWithTimeout returning success or timeouit..
I do think it would be cleaner if exceptionCount was not a passed in incremented reference but a return state, e.g.:
```
if (!attemptThreadStartWithTimeout(....)
    exceptionCount++;
```

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+        CPPUNIT_ASSERT(pool->waitAvailable(100));
+        CPPUNIT_ASSERT(pool->waitAvailable(100));
```
seems a bit pointless to retest same?

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+            for (unsigned threadId = 1; threadId <= 200; threadId++)
+            {
+                threadName.appendf("Thread%u", threadId);
+                if (threadId < 101)
```
personal: threadId <= 100 seems clearer

jakesmith commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+            CPPUNIT_ASSERT_EQUAL(0U, threadCompletedCount.load());
+
+            // Wait for 100 threads to complete, leaving the pool at capacity
+            while (pool->runningCount() > 100)
```
this isn't testing if 100 are complete... ? and 100 should be running almost immediately.

jakesmith reviewed (changes requested):
@streeterd - please see comments.

I think some of the condtions are a bit iffy, and the duration thresholds seem a bit arbitrary (may be related to the GH failure?).

It was quite difficult to review because there's a lot here and combined with code moving. It would help if moved code was a separate commit that reviewer could look at in isolation.
And it looks like some of the testing is slightly redundant/duplicated and could be removed.

It's tempting to go for coverage, but try to avoid duplication, repetition, as it makes the code base bigger, it's harder to review and it's harder to revisit and extend.

streeterd commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+    }
+
+    // Helper method to attempt starting a thread with timeout and count exceptions
+    void attemptThreadStartWithTimeout(ThreadParams *threadParams, unsigned timeoutMs, const char *expectedPoolName, unsigned &exceptionCount,
```
The error string is coming from an exception thrown in `CThreadPool::_start` :

```
            CriticalBlock block(crit);
            if (timedout)
            {
                if (!availsem.wait(0)) // make sure take allocated sem if has become available
                {
                    if (noBlock || timeout > 0)
                        throw MakeStringException(0, "No threads available in pool %s", poolname.get());
                    IWARNLOG("Pool limit exceeded for %s", poolname.get());
                }
            }
```

jakesmith replied:
I'm talking about your new exception message on line 716

streeterd commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        unsigned thread3ExceptionCount = 0;
+        attemptThreadStartWithTimeout(&params, UINT_MAX, "TightlyBoundTestPool", thread3ExceptionCount, false);
+        duration = measureThreadStartDuration(beforeStart, 1);
+        CPPUNIT_ASSERT(duration.count() >= 190);
```
This is a recurring issue within the tests.
There is a discrepancy between what you would expect and the reality of timing.
I suspect it is down to scheduling, measurement precision thread execution timing and environments.

streeterd commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        pool->start(&params, params.name);
+        // Wait for both threads to signal that they've started
+        auto duration = measureThreadStartDuration(beforeStart, 2);
+        CPPUNIT_ASSERT(duration.count() < 1000);
```
Implemented in d6d9a88239923b86e59888dc9d9cc4fefc065dd3

streeterd commented on testing/unittests/jthreadpooltests.cpp:
```diff
@@ -0,0 +1,943 @@
+        params = {lifespan200milliseconds, "Thread3"};
+        beforeStart = std::chrono::high_resolution_clock::now();
+        unsigned thread3ExceptionCount = 0;
+        attemptThreadStartWithTimeout(&params, UINT_MAX, "TightlyBoundTestPool", thread3ExceptionCount, false);
```
Fixed in d6d9a88239923b86e59888dc9d9cc4fefc065dd3

streeterd committed:
Re-code tests to a table format for ease of use.

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>

streeterd committed:
HPCC-34245 TABLE WIP

Signed-off-by: Dave Streeter <dave.streeter@lexisnexisrisk.com>
